### DISCUSSION TITLE
RuntimeError: Peer-to-peer device memory access is not supported

### DISCUSSION THREAD

**Discussion by intelligi123** [NVIDIA/cuQuantum#86] [Q&A] [answered]
I am running Nvidia cuquantum appliance on a server having 4 GPUs (A40) available through a virtual machine made using ESXi.
While running ghz.py example with --gpu > 1, it gave above error:

```
(cuquantum-23.06) cuquantum@da1628726460:~/examples$ python ghz.py --ngpus 3
Traceback (most recent call last):
  File "/home/cuquantum/examples/ghz.py", line 72, in <module>
    main(nqubits=args.nqubits, nrepetitions=args.nsamples, ngpus=args.ngpus)
  File "/home/cuquantum/examples/ghz.py", line 62, in main
    simulator = qsimcirq.QSimSimulator(qsim_options=qsim_options)
  File "/home/cuquantum/conda/envs/cuquantum-23.06/lib/python3.9/site-packages/qsimcirq/qsim_simulator.py", line 262, in __init__
    qsim_mgpu.qsim_initialize_devices(gpu_mode)
RuntimeError: Peer-to-peer device memory access is not supported
```
When I increase --qubits option to see gpu usage only one of the four gpus show 100% usage while rests are not being used despite the fact that container was started with all gpus 

`sudo docker run --gpus all -it --rm nvcr.io/nvidia/cuquantum-appliance:23.06`

output of nvidia-smi while ghz.py running with nqubits=32 is

```
(cuquantum-23.06) cuquantum@da1628726460:~$ nvidia-smi
Tue Sep 12 11:17:22 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.199.02   Driver Version: 470.199.02   CUDA Version: 11.8     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A40          Off  | 00000000:03:00.0 Off |                  Off |
|  0%   43C    P0   185W / 300W |  33510MiB / 48685MiB |    100%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A40          Off  | 00000000:03:01.0 Off |                    0 |
|  0%   31C    P8    21W / 300W |      7MiB / 45634MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  NVIDIA A40          Off  | 00000000:03:02.0 Off |                    0 |
|  0%   31C    P8    21W / 300W |      7MiB / 45634MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  NVIDIA A40          Off  | 00000000:03:03.0 Off |                    0 |
|  0%   30C    P8    21W / 300W |      7MiB / 45634MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```

Does this mean container has no more access than 1 gpu? How can I make sure that all GPUs are being used to their full potential.

Thank you.

**Comment by mtjrider** [comment]
Hi @intelligi123 

Thanks for posting!
Can you share the output of the following commands?
```
nvidia-smi topo -m
nvidia-smi topo -p2p rw
```

The Cirq frontend in the cuQuantum Appliance requires all GPUs be peer accessible. The `ghz.py` example uses Cirq. This frontend is limited to a single node. 

The qiskit frontend (using cusvaer as a backend) relaxes this requirement to enable multi-node deployment.

**Reply by intelligi123** [reply]
Yes sure, below are the required command outputs, I am also interested in qiskit but couldn;t find any example which takes number of gpus as input to check. Can you please share any examples which uses multiple GPUs/nodes in qiskit for tasks distribution?

(```
cuquantum-23.06) cuquantum@da1628726460:~/examples$ nvidia-smi topo -m
	GPU0	GPU1	GPU2	GPU3	CPU Affinity	NUMA Affinity
GPU0	 X 	PHB	PHB	PHB	0-7		N/A
GPU1	PHB	 X 	PHB	PHB	0-7		N/A
GPU2	PHB	PHB	 X 	PHB	0-7		N/A
GPU3	PHB	PHB	PHB	 X 	0-7		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks
```
and 
```
(cuquantum-23.06) cuquantum@da1628726460:~/examples$ nvidia-smi topo -p2p rw
 	GPU0	GPU1	GPU2	GPU3	
 GPU0	X	CNS	CNS	CNS	
 GPU1	CNS	X	CNS	CNS	
 GPU2	CNS	CNS	X	CNS	
 GPU3	CNS	CNS	CNS	X	

Legend:

  X    = Self
  OK   = Status Ok
  CNS  = Chipset not supported
  GNS  = GPU not supported
  TNS  = Topology not supported
  NS   = Not supported
  U    = Unknown
```

**Reply by mtjrider** [reply, marked as answer]
Thanks! `CNS`, meaning `chipset not supported`, would explain the error you posted earlier.

Regarding your question seeking additional examples: please have a look at this [documentation page](https://docs.nvidia.com/cuda/cuquantum/latest/appliance/cusvaer.html#using-cusvaer). Notably:

```]
$ cat ghz_cusvaer.py

from qiskit import QuantumCircuit, transpile
from cusvaer.backends import StatevectorSimulator

def create_ghz_circuit(n_qubits):
    ghz = QuantumCircuit(n_qubits)
    ghz.h(0)
    for qubit in range(n_qubits - 1):
        ghz.cx(qubit, qubit + 1)
    ghz.measure_all()
    return ghz

circuit = create_ghz_circuit(20)

simulator = StatevectorSimulator()
simulator.set_options(precision='double')
circuit = transpile(circuit, simulator)
job = simulator.run(circuit, shots=1024)
result = job.result()

if result.mpi_rank == 0:
    print(result.get_counts())
```

Running this example with: `mpirun -n NUM_GPUS ghz_cusvaer.py`
